source/device.hpp

```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include <set>
+#include <unordered_map>
+
+namespace sim {
+
+class Link;
+
+enum DeviceType { SWITCH, SENDER, RECEIVER };
+
+class Device {
+public:
+    Device(DeviceType a_type);
+    virtual ~Device();
+    virtual void process();
+
+private:
```

Comment:
Dont we need protected modifier here?
```suggestion
protected:
```